kong/runloop/handler.lua

```diff
@@ -365,6 +372,190 @@ local function register_events()
+    else
+      log(level, message)
+    end
+  end
+
+
+  local function safe_rebuild(name, callback, version)
+    local max_attempts = 6
+    local success = false
+
+    local action = version == "init" and "initializing" or "rebuilding"
+
+    for attempt = 1, max_attempts do
+      log_with_worker_id(DEBUG,
+        fmt("%s %s. Attempt %d/%d", action, name, attempt, max_attempts))
+      local pok, ok, err = pcall(callback, version)
+      if pok and ok then
+        log_with_worker_id(DEBUG, fmt("%s %s succeeded", action, name))
+        success = true
+        break
+      end
+
+      log_with_worker_id(NOTICE, fmt("%s %s failed: %s", action, name, ok or err))
+      sleep(0.01 * attempt * attempt)
+    end
+
+    if not success then
+      return nil, fmt("%s %s failed %s times, aborting",
+                      action, name, max_attempts)
+    end
+
+    return success
+  end
+
+
+  local function rebuild_timer(premature, name, callback, version, semaphore)
+    if premature then
+      release_semaphore(semaphore)
+      return
+    end
+    local ok, err = safe_rebuild(name, callback, version)
+    if not ok then
+      log(CRIT, err)
+    end
+    release_semaphore(semaphore)
+  end
+
+
+  local function rebuild_async(name, callback, version, semaphore)
+    local ok, err = timer_at(0, rebuild_timer, name, callback, version, semaphore)
+    if not ok then
+      return nil, "could not create rebuild timer: " .. err
+    end
+
+    return true
+  end
+
+
+
+  -- @param name "router" or "plugins_iterator"
+  -- @param callback A function that will update either the router or plugins_iterator
+  -- @param version target version
+  -- @param semaphore router_semaphore or plugins_iterator_semaphore
+  -- @param wait how much time to wait when trying to acquire the semaphore.
+  -- @returns true if safe_rebuild was either successfully executed synchronously,
+  -- enqueued via async timer, or not needed (because current_version == target).
+  -- nil otherwise (safe_rebuild was neither called successfully nor enqueued).
+  -- Notice that this function might return true *even when it fails to acquire
+  -- the semaphore or invoke the async callback* - it will try to rebuild
+  -- synchronously when a wait is specified and the semaphore isn't acquired
+  -- or the timer-based call fails.
+  -- @returns error message as a second return value in case of failure
+  rebuild = function(name, callback, version, semaphore, wait)
+    local current_version, err = get_version(name)
+    if not current_version then
+      return nil, err
+    end
+    if current_version == version then
+      return true
+    end
+
+    local ok = acquire_semaphore(semaphore, wait)
+    if not ok then
+      if wait and wait > 0 then
+        log(NOTICE, "could not acquire semaphore, attempting sync rebuild")
+        local ok, err = safe_rebuild(name, callback, current_version)
+        if not ok then
+          return nil, err
```

Comment:
Again here it is marked as fatal for the request. I question that a bit as I think it is better to serve request with `stale` router/plugins than refuse to serve request at all. This affects usually only one request. If Kong never is going to make a new rebuild even on other requests, you just got a lot of lines `CRIT` lines in your log, but it will still allow you to proxy requests with old router, until someone acts upon it. With making these fatal and rebuild never succeeding, you practically make all request stop working. Think about this happening middle of a night. That was the original reason that I made these to log with CRIT/ERR.

Of course this is debatable. Yours is more correct in a way that Kong immediately refuses to serve requests with stale router/plugins. But I would say that mine is more resilient on error and allows at least `99.99%` of the traffic still be processed 100% correct.

Though I am not sure if you handled ^ elsewhere, making this comment obsolete, but you got the idea?

Reply:
I have changed how this works now (by using the `get_updated_plugins_iterator` field, which logs and serves the old version if needed).